js/modules/loki_public_chat_api.js

```diff
@@ -72,13 +79,102 @@ class LokiPublicServerAPI {
     this.channels.splice(i, 1);
     thisChannel.stopPolling = true;
   }
+
+  async getServerToken() {
+    let token = await Signal.Data.getPublicServerTokenByServerName(this.server);
+    if (!token) {
+      token = await this.getNewToken();
+      if (token) {
+        await Signal.Data.savePublicServerToken({
+          server: this.server,
+          token,
+        });
+      }
+    }
+    return token;
+  }
+
+  async getNewToken() {
+    if (!this.tokenPending) {
+      this.tokenPending = true;
+      this.tokenPromise = new Promise(async res => {
+        const token = await this.requestToken();
+        if (!token) {
+          res(null);
+          return;
+        }
+        const registered = await this.submitToken(token);
+        if (!registered) {
+          res(null);
+          return;
+        }
+        res(token);
+      });
+    }
+    const token = await this.tokenPromise;
+    this.tokenPending = false;
+    return token;
+  }
+
+  async requestToken() {
+    const url = new URL(`${this.baseServerUrl}/loki/v1/get_challenge`);
+    const params = {
+      pubKey: this.chatAPI.ourKey,
+    };
+    url.search = new URLSearchParams(params);
+
+    let res;
+    try {
+      res = await nodeFetch(url);
+    } catch (e) {
+      return null;
+    }
+    if (!res.ok) {
+      return null;
+    }
+    const body = await res.json();
+    const { cipherText64, serverPubKey64 } = body;
+    const token = await libloki.crypto.decryptToken(
+      cipherText64,
+      serverPubKey64
+    );
+    return token;
+  }
+
+  async submitToken(token) {
+    const options = {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        pubKey: this.chatAPI.ourKey,
+        token,
+      }),
+    };
+
+    let res;
+    let success = true;
+    try {
+      res = await nodeFetch(
+        `${this.baseServerUrl}/loki/v1/submit_challenge`,
+        options
+      );
+      success = res.ok;
+    } catch (e) {
+      return false;
+    }
+    return success;
```

Comment:
You should be able to declare `res` and return the value directly from within the `try` block:
```js
try {
  const res = ...
  return res.ok;
}
```

Reply:
Good call